### PR TITLE
REF: separate numpy-specific part of can_hold_element

### PR DESCRIPTION
**Pull request by jbrockmendel**
By returning the value instead of just a bool, we open the option to return the _casted_ value, which is what we'll do in the follow-up, which will let us de-duplicate NumericIndex._validate_fill_value.  By separating out the np-specific parts, we'll be able to improve perf in a couple of places.

### CODE REVIEW

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -2185,80 +2185,129 @@ def can_hold_element(arr: ArrayLike, element: Any) -> bool:
...
 
-    if dtype == _dtype_obj:
+    try:
+        np_can_hold_element(dtype, element)
         return True
+    except (TypeError, ValueError):
+        return False
+
+
+def np_can_hold_element(dtype: np.dtype, element: Any):
```
Comment (jreback) [inline pandas/core/dtypes/cast.py:2195]: can you add a return Any

File: pandas/core/dtypes/cast.py [hunk 2]
```diff
@@ -2185,80 +2185,129 @@ def can_hold_element(arr: ArrayLike, element: Any) -> bool:
...
+    """
+    Raise if we cannot losslessly set this element into an ndarray with this dtype.
+
+    Specifically about places where we disagree with numpy.  i.e. there are
+    cases where numpy will raise in doing the setitem that we do not check
+    for here, e.g. setting str "X" into a numeric ndarray.
+
+    Returns the element, potentially cast to the dtype.
+
```
Comment (jreback) [inline pandas/core/dtypes/cast.py:2204]: add a Returns section

**Review by jreback — CHANGES_REQUESTED** [review]
would it be better to use `None` as the return value rathe than using exceptions as control flow ? (e.g. performance)

**Comment by jbrockmendel** [comment]
> would it be better to use None as the return value rathe than using exceptions as control flow ? (e.g. performance)

That becomes ambiguous with if the actual element is None.

There _are_ some performance concerns i have with this function, but try/except overhead is not among them.

**Comment by jreback** [comment]
> There are some performance concerns i have with this function, but try/except overhead is not among them.

oh ok.

**Review by jreback — APPROVED** [review]

**Pull request merged**